[TOPIC-GPIO] gpio: Update mcux lpc driver to use new gpio api

agansari opened the pull request:
Updates the mcux lpc driver and LPC boards to use new device tree compatible gpio configuration flags. Implements new port get/set/clear/toggle and pin_interrupt_configure functions recently added to the gpio api.

Driver still supports port acces on some features. To be removed later.

Interrupts feature enabled in this pull in order to have a wider test array.
Ported from #16589, will be closed once this is merged.

Tested with:
 - samples/basic/blinky (pass - lpc54 and lpc55)
 - samples/basic/button (pass - lpc54 and lpc55)
 - tests/drivers/gpio/gpio_api_1pin (pass - lpc54)
 - tests/drivers/gpio/gpio_basic_api (pass - lpc54)

On boards:
 - lpcxpresso54114_m4
 - lpcxpresso55s69_cpu0

Tests in tests/drivers/ pass on lpc55s69 pass if run in debug mode.

agansari committed:
boards: lpcxpresso55s69 pinmux macros refactor

Board is refactored to use DTS generated value, not use
magic numbers.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
dts: NXP LPC gpio updates

NXP's LPC family of MCU's GPIOs parameters is udated.
Boards LPC54xxx and LPC55xxx have updated values according
pin and interrupt layout.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
soc: enable PINT device at LPC SoC boot

PINT device is enabled when SoC is booting up. Applies to LPC54xxx and
LPC55xxx families.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
gpio: Update mcux lpc driver to use new gpio api

Updates the mcux lpc driver and all associated boards to use new
device tree compatible gpio configuration flags. Implements new port
get/set/clear/toggle functions recently added to the gpio api.
ISR functions to be added later.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
west.yaml: hal_nxp updated to latest commit

Use the latest commit available in hal_nxp repository.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
soc: LPC55xxx clock comment fix

SoC initialization had an incorrect comment regarding system clock.
Corrected from 48Mhz -> 96Mhz.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
tests: gpio_basic_api fix bad message

Pull-up replaced by pull-down in particular test_gpio_port.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
tests: lpc devices test pins enablement

LPC54114 to use D0 and D1 pinout.
LPC55S69 to use A0 and A1 pinout.
Pins enabled to be used as GPIO for 2 pin test gpio_basic_api.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
tests: add LPC boards to gpio_basic_api

LPC54114 to use D0 and D1 pinout.
LPC55S69 to use A0 and A1 pinout.
2 pin test gpio_basic_api uses pins set up in boars' pinmux.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
soc: define gpio pull-down for LPC54114

Add define that maps to IOCON register PULL-DOWN bit.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
dts: lpc devices allocate all gpio interrupts

Allocate all 8 PINT interrupts to ports 0 and 1, allocate 4 to each.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
drivers: gpio_mcux_lpc GPIO interrupts

LPC GPIO architecture uses multiple devices.
GPIO input is routed via INPUTMUX to the PINT
device which roots the interrupt to NVIC.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

mnkp commented on drivers/gpio/gpio_mcux_lpc.c:
```diff
@@ -20,82 +19,119 @@
+	/* pin association with PINT id */
+	pint_pin_int_t pint_id[32];
+	/* interrupt type */
+	pint_pin_enable_t int_mode[32];
```
I didn't dig into interrupt architecture much, however the `int_mode` doesn't seem to be used. There is one line which assigns a value to the field but there doesn't seem to be one that reads it. We are using here 32 bytes (or rather sizeof(pint_pin_enable_t)) of static RAM per port. We shouldn't if we don't have to.

agansari replied:
Will be removed, it's a remainder from porting to the new api and how interrupt is separate from general config.

mnkp reviewed (changes requested):
LGTM with minor (but blocking) comments.

agansari commented on boards/arm/lpcxpresso54114/pinmux.c:
```diff
@@ -90,6 +89,27 @@ static int lpcxpresso_54114_pinmux_init(struct device *dev)
 			);
 	pinmux_pin_set(port0,  4, port0_pin4_config);
+
+#ifdef DT_INST_0_TEST_GPIO_BASIC_API
```
Agree that test specific stuff should go inside the test's code, but i find it messy.
I find using board's pinmux to be a cleaner approach whether it's an application, sample or test.
I also mentioned this in https://github.com/zephyrproject-rtos/zephyr/pull/19247#discussion_r332057888

mnkp replied:
> I find using board's pinmux to be a cleaner approach

I agree, but at the moment all other drivers placed the pinmux configuration directly in the testcase. We know that what we're currently doing is not a clean approach but there was no time to do it properly. This will be fixed. It will be easier to work on the fix if all the drivers use the same approach for now. The same applies to `boards/arm/lpcxpresso55s69/pinmux.c`.

mnkp replied:
Actually only frdm_k64f board has its pinmux configuration done in the testcase at the moment. Other PRs are not yet merged. Doing pinmux configuration directly in the board file makes sense but it really shouldn't be testcase specific. If we follow this road the board files will quickly become unmaintainable. Unfortunately, the approach to running tests on hardware is not fully worked out at the moment.

carlescufi commented:
@agansari 
The topic branch has been rebased. Could you please rebase this PR against it?

mnkp commented on drivers/gpio/gpio_mcux_lpc.c:
```diff
@@ -20,82 +19,110 @@
 };
 
 struct gpio_mcux_lpc_data {
-	/* gpio_driver_data needs to be first */
```
I would keep the comment, it's present in all other GPIO drivers.

mnkp commented:
It looks like something went wrong during the rebase process. There are some commits in this PR which shouldn't be here. It will probably require some manual clean up.

agansari commented:
> It looks like something went wrong during the rebase process. There are some commits in this PR which shouldn't be here. It will probably require some manual clean up.

@mnkp I rebased and moved the pin configuration to the test.

mnkp commented on drivers/gpio/gpio_mcux_lpc.c:
```diff
@@ -111,49 +139,396 @@ static int gpio_mcux_lpc_read(struct device *dev,
+
+	/* Pin used as interrupt is not set as output*/
+	if ((mode & GPIO_INT_ENABLE) &&
+		((gpio_base->DIR[port] >> pin) ^ 0)) {
```
That doesn't look correct. Shouldn't it be: `((gpio_base->DIR[port] & BIT(pin)) == 0)`? We're testing if pin is an input.

mnkp commented on drivers/gpio/gpio_mcux_lpc.c:
```diff
@@ -111,49 +139,396 @@ static int gpio_mcux_lpc_read(struct device *dev,
+	u32_t isr;
+	u32_t pint_idx;
+
+	/* Pin used as interrupt is not set as output*/
```
input?

agansari replied:
@mnkp nope, interrupts on output don't work.
I rephrased to: "/* Ensure pin used as interrupt is set as input*/"

mnkp commented on drivers/gpio/gpio_mcux_lpc.c:
```diff
@@ -111,49 +139,396 @@ static int gpio_mcux_lpc_read(struct device *dev,
+	/* Ensure pin used as interrupt is set as input*/
+	if ((mode & GPIO_INT_ENABLE) &&
+		((gpio_base->DIR[port] & BIT(pin)) != 0)) {
+		return -EINVAL;
```
Other drivers return `-ENOTSUP`. We are saying that the flag configuration is valid, however is not supported by the hardware.

agansari replied:
Agree

mnkp commented on drivers/gpio/gpio_mcux_lpc.c:
```diff
@@ -111,49 +139,396 @@ static int gpio_mcux_lpc_read(struct device *dev,
+		isr = get_free_isr(data);
+		if (isr == -EINVAL) {
+			/* Didn't find any free interrupt in this port */
+			return -EINVAL;
```
Other drivers return `-EBUSY` meaning "Interrupt line required to configure pin interrupt is already in use."

agansari replied:
@mnkp done

mnkp reviewed (commented):
Thanks for the update! Sorry, I overlooked two other minor issues. However, we should fix them to make sure the testcases can pass.

agansari commented:
failing test is qemu base, not related to this pull

mnkp commented:
It looks like this PR needs to be rebased after recent merges to `topic-gpio` branch.

agansari commented:
> It looks like this PR needs to be rebased after recent merges to `topic-gpio` branch.

@mnkp rebase and shippable passed

MaureenHelm commented on boards/arm/lpcxpresso55s69/pinmux.c:
```diff
@@ -8,6 +8,7 @@
 #include <fsl_common.h>
 #include <fsl_iocon.h>
 #include <soc.h>
+#include <LPC55S69_cm33_core0.h>
```
Why are you adding this include?

agansari replied:
i forgot to clean this up

MaureenHelm commented on dts/bindings/gpio/nxp,lpc-gpio.yaml:
```diff
@@ -0,0 +1,22 @@
+title: LPC GPIO
```
Please add a license/copyright header

MaureenHelm commented on drivers/gpio/gpio_mcux_lpc.c:
```diff
@@ -44,27 +43,34 @@ static int gpio_mcux_lpc_configure(struct device *dev,
-		} else {
-			gpio_base->SET[config->port_no] = BIT(pin);
-			gpio_base->DIR[config->port_no] |= BIT(pin);
+		WRITE_BIT(data->general.invert, pin,
```
You shouldn't need to handle the invert field. This is now handled in `include/drivers/gpio.h`

agansari replied:
this was removed in a previous commit

MaureenHelm commented on drivers/serial/usart_mcux_lpc.c:
```diff
@@ -290,17 +292,17 @@ static void usart_mcux_lpc_config_func_0(struct device *dev);
-	.baud_rate = DT_USART_MCUX_LPC_0_BAUD_RATE,
+	.base = (USART_Type *)DT_NXP_LPC_USART_USART_0_BASE_ADDRESS,
+	.clock_source = kCLOCK_FlexComm0,
+	.baud_rate = DT_NXP_LPC_USART_USART_0_CURRENT_SPEED,
```
Why are there USART refactoring patches in a GPIO PR?

agansari replied:
will drop commit and move to a separate pull

MaureenHelm replied:
Thanks

agansari commented:
@MaureenHelm can you review this again?
